mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,273 @@
+        labeled_logs_pattern: str,
+        unlabeled_logs_path: str,
+        output_folder: str,
+    ):
+        """
+        Args:
+            app_path (str): Path to MindMeld application
+            train_pattern (str): Regex pattern to match train files. For example, ".*train.*.txt"
+            test_pattern (str): Regex pattern to match test files. For example, ".*test.*.txt"
+            load (bool): Whether to load pickled queries from a local folder
+            save (bool): Whether to save queries as a local pickle file
+            train_seed_pct (float): Percentage of training data to use as the initial seed
+            n_classifiers (int): Number of classifiers to be used by a subset of heuristics
+            n_epochs (int): Number of epochs to run training
+            batch_size (int): Number of queries to select at each iteration
+            training_level (str): The hierarchy level to train ("domain" or "intent")
+            training_strategies (List[str]): List of strategies to use for training
+            selection_strategy (str): Single strategy to use for log selection
+            save_sampled_queries (bool): Whether to save the queries sampled at each iteration
+            log_usage_pct (float): Percentage of the log data to use for selection
+            labeled_logs_pattern (str): Pattern to obtain logs already labeled in a MindMeld app
+            unlabeled_logs_path (str): Path a logs text file with unlabeled queries
+            output_folder (str): Folder to store active learning results.
+        """
+        self.app_path = app_path
+        self.train_pattern = train_pattern
+        self.test_pattern = test_pattern
+        self.load = load
+        self.save = save
+        self.train_seed_pct = train_seed_pct
+        self.n_classifiers = n_classifiers
+        self.n_epochs = n_epochs
+        self.batch_size = batch_size
+        self.training_level = training_level
+        self.training_strategies = training_strategies
+        self.selection_strategy = selection_strategy
+        self.save_sampled_queries = save_sampled_queries
+        self.log_usage_pct = log_usage_pct
+        self.labeled_logs_pattern = labeled_logs_pattern
+        self.unlabeled_logs_path = unlabeled_logs_path
+        self.output_folder = output_folder
+
+        self.results_manager = ResultsManager(output_folder)
+        self.mindmeld_al_classifier = self._get_mindmeld_al_classifier()
+        self.init_train_data_bucket = None
+        self.data_bucket = None
+
+    def _get_mindmeld_al_classifier(self):
+        """ Creates an instance of a MindMeld Active Learning Classifier. """
+        return MindMeldALClassifier(
+            self.app_path, self.training_level, self.n_classifiers
+        )
+
+    @property
+    def num_iterations(self) -> int:
+        """
+        Returns:
+            num_iterations (int): Number of iterations needed for training.
+        """
+        return 1 + math.ceil(
```

Comment:
Could you add a comment or var here about the "1 + .."

Reply:
Added `# An additional iteration is added to save training data after the last sampling round.`